tg/models.py

```diff
@@ -340,63 +342,55 @@ def jump_bottom(self, chat_id: int) -> bool:
 
     def prev_msg(self, chat_id: int, step: int = 1) -> bool:
         new_idx = self.current_msgs[chat_id] + step
-        if new_idx < len(self.msgs[chat_id]):
+        if new_idx < len(self.msg_ids[chat_id]):
             self.current_msgs[chat_id] = new_idx
             return True
         return False
 
     def get_message(self, chat_id: int, msg_id: int) -> Optional[Dict]:
         if msg_id in self.not_found:
             return None
-        if index := self.msg_idx[chat_id].get(msg_id):
-            return self.msgs[chat_id][index]
-        # we are not storing any out of ordres old msgs
-        # just fetching them on demand
+        if msg := self.msgs[chat_id].get(msg_id):
+            return msg
         result = self.tg.get_message(chat_id, msg_id)
         result.wait()
         if result.error:
             self.not_found.add(msg_id)
             return None
         return result.update
 
-    def remove_messages(self, chat_id: int, msg_idx: List[int]) -> None:
-        log.info(f"removing msg {msg_idx=}")
-        self.msgs[chat_id] = [
-            m for m in self.msgs[chat_id] if m["id"] not in msg_idx
-        ]
-        self.msg_idx[chat_id] = {
-            msg["id"]: i for i, msg in enumerate(self.msgs[chat_id])
-        }
+    def remove_messages(self, chat_id: int, msg_ids: List[int]) -> None:
+        log.info(f"removing msg {msg_ids=}")
+        for msg_id in msg_ids:
+            self.msg_ids[chat_id].remove(msg_id)
+            self.msgs[chat_id].pop(msg_id, None)
 
     def add_message(self, chat_id: int, msg: Dict[str, Any]) -> None:
         log.info(f"adding {msg=}")
-        self.msgs[chat_id] = sorted(
-            self.msgs[chat_id] + [msg], key=lambda d: d["id"], reverse=True,
-        )
-        self.msg_idx[chat_id] = {
-            msg["id"]: i for i, msg in enumerate(self.msgs[chat_id])
-        }
+        msg_id = msg["id"]
+        self.msgs[chat_id][msg_id] = msg
+        self.msg_ids[chat_id].append(msg_id)
+        self.msg_ids[chat_id].sort(reverse=True)
```

Comment:
we can remove redundant sort operations with:
``` if msg_id < self.msg_ids[chat_id][1]: self.msg_ids[chat_id].sort(reverse=True)```

Reply:
Did you mean `if msg_id < self.msg_ids[chat_id][-2]:`?
Cause when sorted, the biggest msg id is in the 0 index.
Or did I get something wrong?